### PR TITLE
doc: build: kconfig: subtlety of multiple defs without deps

### PR DESCRIPTION
**Pull request by JordanYates**
Document a subtlety of the "Dependencies of multiple symbol definitions are ORed together" behaviour when applied to nodes without any explicit `depends on X` listed.

This behaviour can be tested by adding the following to `Kconfig.zephyr` and adding `CONFIG_A=y` to any `prj.conf`.
```
config A
    bool "Test Symbol A"

config B
    bool "Test Symbol B"

config TEST_SYMBOL
    bool "Test symbol"
    depends on B

config TEST_SYMBOL
    default y if A
```
`CONFIG_TEST_SYMBOL` will be enabled despite `CONFIG_B` being disabled.

### CODE REVIEW

**Comment by JordanYates** [comment]
Personally I would prefer this to be fixed instead of documented, but in a few hours of playing around with `kconfiglib.py` I'm still mostly at a loss as to how to do that. IMO the problem is with
```
direct_dep:
      The direct ('depends on') dependencies for the symbol, or self.kconfig.y
      if there are no direct dependencies.
```
Using `self.kconfig.y` as the default overrides any other dependencies when they are ORed together, but changing the default to `self.kconfig.n` breaks the Kconfig process in ways that I don't understand.

**Comment by carlocaione** [comment]
I encountered this issue a couple of days ago when working on the cache rework. We have:
```
config CACHE_MANAGEMENT
    bool "Cache management features"
    depends on DCACHE || ICACHE
```
And a SoC having:
```
config CACHE_MANAGEMENT
    default y
```
will set it even when `DCACHE` or `ICACHE` are not enabled.

File: doc/build/kconfig/setting.rst [hunk 1]
```diff
@@ -259,9 +259,24 @@ symbol properties, so the above ``default`` is equivalent to
       	...
       	depends on DEP2
 
+   Symbols without any listed dependencies are treated as having a dependency
```
Comment (tejlmand) [inline doc/build/kconfig/setting.rst:262]: what is wrong with the `warning` above ?
image

File: doc/build/kconfig/setting.rst [hunk 2]
```diff
@@ -259,9 +259,24 @@ symbol properties, so the above ``default`` is equivalent to
...
+      	depends on DEP1
+
+      config FOO
+      	...
+      	default y if BAR
+
    When making changes to :file:`Kconfig.defconfig` files, always check the
    symbol's direct dependencies in one of the :ref:`interactive configuration
-   interfaces <menuconfig>` afterwards. It is often necessary to repeat
+   interfaces <menuconfig>` afterwards. It is **required** to repeat
```
Comment (tejlmand) [inline doc/build/kconfig/setting.rst:279]: not accepted.

Often a downstream project actually want to weaken a symbol, like this:
```
# Kconfig.zephyr

config FOO
    bool "Foo"
    depends on BAR
```

but downstream has an alternative feature:
```
# Kconfig.downstream

config FOO
    depends on DOWN_BAR
```

and that's a perfectly valid use-case.
But stating required can mislead the reader into believing it is required to duplicate all dependencies.

Also in a situation where the repeated symbol is guarded by `if`, then it is also not necessary to repeat the `depends on`.

Like:
```
# Kconfig.A

config FOO
    depends on BAR
```
and then having Kconfig.B which sources a Kconfig.C which changes the default, like:
```
# Kconfig.B

if BAR

source Kconfig.C

endif
```
and then

```
# Kconfig.C

config FOO
    default y
```
in Kconfig.C there is no need for repeating the `depends on` due to the `if BAR` before the source line.

**Review by tejlmand — CHANGES_REQUESTED** [review]
I believe this is already well described.

It's even in a big yellow warning box.

**Comment by tejlmand** [comment]
> Personally I would prefer this to be fixed instead of documented

First, then how will you allow a symbol to be weakened when there is an alternative symbol providing similar option ?
Secondly, the Kconfig language is not defined by Zephyr, but by `kernel.org`, kconfiglib is just implementing the existing language.

File: doc/build/kconfig/setting.rst (hunk 1 above)
Comment (tejlmand) [reply doc/build/kconfig/setting.rst:262]: maybe you should look at the definition of `depends on`:
> Dependencies are applied to all other options within this menu entry (which also accept an “if” expression)

Ref: https://www.kernel.org/doc/html/latest/kbuild/kconfig-language.html

which basically means that the new example you are creating:
```
config FOO
    ...
    depends on DEP1
```
is identical to:
```
config FOO
    ... if DEP1
```
which suddenly makes it very clear why Kconfig behave the way it does, cause now you have:
```
config FOO
    ... if DEP1
```
and:
```
 config FOO
      	...
      	default y if BAR
```

Comment (JordanYates) [reply doc/build/kconfig/setting.rst:262]: The warning above makes no mention of what the behaviour is when an entry doesn't have a `depends on`. Regardless of what the resolution is here, there should be some mention of this (as proven by @carlocaione running into the same behaviour and considering it unintuitive).

Your example doesn't really clear it up for me, why should the concrete example of
```
config CACHE_MANAGEMENT
    bool "Cache management features"
    depends on DCACHE || ICACHE

config CACHE_MANAGEMENT
    default y
```
Be equivalent to:
```
config CACHE_MANAGEMENT
    bool "Cache management features"
    default y
```

Also as far as I can tell the linked reference makes no mention of how multiple symbols are handled, beyond:
```
A config option can be defined multiple times with the same name, but every definition can have only a single input prompt and the type must not conflict.
```

**Comment by JordanYates** [comment]
> > Personally I would prefer this to be fixed instead of documented
> 
> First, then how will you allow a symbol to be weakened when there is an alternative symbol providing similar option ? Secondly, the Kconfig language is not defined by Zephyr, but by `kernel.org`, kconfiglib is just implementing the existing language.

I'm not suggesting that symbol weakening be removed, I'm suggesting that symbols defined without any explicit `depends on`, should not be treated as `depends on y`, and thus removing all other dependencies.

In my head this can be done relatively simply when parsing the symbols by:
1. If the new dependency is a constant (it hasn't been explicitly specified in Kconfig), keep the current dependency
2. If the new dependency is an expression
  a.  If the current dependency is constant, overwrite it
  b. If the current dependency is an expression, OR it

I believe this would avoid nodes without `depends on` removing all other dependencies.

Or if the expectation is that if you want to add a default to a symbol, you should find and duplicate the original dependencies (and manually keep them up to date), then that should also be explicitly called out in the documentation. Personally I want to define a symbol in my downstream module, which when selected applies a bunch of sane defaults for my applications. My own bias speaking, but I would think this is a relatively common desire.
```
config MY_COMPANY_APPS
    bool "Apply defaults I like"

config BT
    default y If MY_COMPANY_APPS
config MCUMGR
    default y if MY_COMPANY_APPS && BT
...
```

File: doc/build/kconfig/setting.rst (hunk 1 above)
Comment (tejlmand) [reply doc/build/kconfig/setting.rst:262]: How about the whole part and example, https://www.kernel.org/doc/html/latest/kbuild/kconfig-language.html:
(Unfortunately I cannot link directly to the description)
> Dependencies are applied to all other options within this menu entry (which also accept an “if” expression), so these two examples are equivalent:
>
> ```
> bool "foo" if BAR
> default y if BAR
> ```
> and:
> ```
> depends on BAR
> bool "foo"
> default y
> ```

So when you write:
```
config CACHE_MANAGEMENT
    bool "Cache management features"
    depends on DCACHE || ICACHE

config CACHE_MANAGEMENT
    default y
``` 

you are actually writing:
```
config CACHE_MANAGEMENT
    bool "Cache management features" if DCACHE || ICACHE

config CACHE_MANAGEMENT
    default y
```

guess it's clearer now that you are actually defining two `CACHE_MANAGEMENT` settings.
One which has a prompt if `DCACHE` or `ICACHE` is enabled.
And a promptless setting which simply defaults `CACHE_MANAGEMENT` unconditionally to `y`.

**Comment by tejlmand** [comment]
> I'm suggesting that symbols defined without any explicit `depends on`, should not be treated as `depends on y`, and thus removing all other dependencies.

what you are suggestion is that Kconfiglib is not following the language specification.
I linked you to the specification of the Kconfig language, which describes how `depends on` works.

**Comment by JordanYates** [comment]
> btw, please note that i'm not against improving the existing description.
> 
> I just find it important that we are precise in how we describe and improve the text.
> 
> Current description focus to much on trying to describe `depends on`, when in fact `depends on` can be thought of as a short hand for several `if <something>`, in exact same way a `<type> "<text>"` is a short hand for
> 
> ```
> <type>
> prompt "<text>"
> ```
> 
> Ref: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#common-kconfig-shorthands

The logic is clearer to me now, thanks for the explanations. RE the docs improvements, I am less concerned about the specifics of `depends on` and more interested in describing the process that should be followed if someone wishes to add a new `default` to a symbol that is already defined by Zephyr.

Please correct me if I'm wrong, but it appears as if the options are:
1. Fork Zephyr, add the `default` to the original symbol definition
2. Define the symbol a second time with the new default, while duplicating any `depends on` from the original symbol (Also considering any if statements surrounding it). Periodically check all symbols defined this way to keep dependencies in sync.

If you agree with the two options above I will have another go at updating the docs, specifically addressing that use case.

**Comment by JordanYates** [comment]
Preferred solution is in #51316

**Pull request closed without merging**